src/kafka-util/src/client.rs

```diff
@@ -35,8 +36,123 @@ pub const DEFAULT_FETCH_METADATA_TIMEOUT: Duration = Duration::from_secs(30);
 /// All code in Materialize that constructs Kafka clients should use this
 /// context or a custom context that delegates the `log` and `error` methods to
 /// this implementation.
-#[derive(Clone)]
-pub struct MzClientContext;
+#[derive(Clone, Default)]
+pub struct MzClientContext {
+    /// The last observed error log, if any.
+    error: Arc<Mutex<Option<MzKafkaError>>>,
+}
+
+impl MzClientContext {
+    /// Retrieves the most recent error since the last invocation of this method, if any. This
+    /// method clears the error and future invocations will return `None` unless a new error is
+    /// logged.
+    pub fn take_err(&self) -> Option<MzKafkaError> {
+        self.error.lock().unwrap().take()
+    }
+
+    fn set_err(&self, msg: &str) {
+        let err = match MzKafkaError::from_str(msg) {
+            Ok(err) => err,
+            Err(()) => {
+                error!(message = msg, "failed to parse kafka error");
+                MzKafkaError::Internal(msg.to_owned())
+            }
+        };
+        match *self.error.lock().unwrap() {
+            // This replacement happens on the basis that internal type errors can be of very low
+            // relevance to our users since librdkafka logs many internal details (e.g
+            // openssl errors) as error logs. Replacing the existing error if it's of the
+            // Internal variant will either result into a new Internal error, or, a known
+            // error variant.
+            Some(ref mut cur @ MzKafkaError::Internal(_)) => *cur = err,
+            // If it's a known error variant we "latch" onto it until it's cleared.
```

Comment:
still, we may have MULTIPLE real errors, and lose data on later ones

Can we make this a vector/map instead, where it can hold one of each type of error max?

Reply:
The reason this is not a `Vec` is that this context is used in a bunch of places and not all of them are wired up to handle errors from the context, so a Vec could be accumulating data for a long time. I considered a middleground where this is a channel and you can drop the receiver to signal that you don't care about errors which would solve the problem but there is nothing that would use the multiple errors atm.

Reply:
I suppose this is a better change than nothing, and we can start with it! Definitely might want to leave a `TODO` to use a bounded buffer/channel of some kind though!

Reply:
I implemented the channel! It wasn't that complicated in the end